dlib/matrix/kiss_fft.h

```diff
@@ -0,0 +1,733 @@
+        inline kiss_fftr_state<T>::kiss_fftr_state(const plan_key& key)
+        {
+            if (key.dims[0] & 1)
+                throw std::runtime_error("real FFT must have even dimension");
+            
+            const int nfft = key.dims[0] / 2;
+            substate = kiss_fft_state<T>(plan_key({nfft}, key.is_inverse));
+            super_twiddles.resize(nfft/2);
+
+            for (size_t i = 0 ; i < super_twiddles.size() ; ++i) 
+            {
+                double phase = -3.141592653589793238462643383279502884197169399 * ((double) (i+1) / nfft + .5);
+                if (key.is_inverse)
+                    phase *= -1;
+                super_twiddles[i] = std::polar(1.0, phase);
+            }
+        }
+
+        template<typename T>
+        void kiss_fftr(const kiss_fftr_state<T>& plan, const T* timedata, std::complex<T>* freqdata)
+        {
+            if (plan.substate.inverse)
+                throw std::runtime_error("bad fftr plan : need a forward plan. This is an inverse plan");
+
+            const int nfft_h = plan.substate.nfft; //recall that the FFT size is actually half the original requested FFT size, i.e. the size of timedata
+
+            /*perform the parallel fft of two real signals packed in real,imag*/
+            std::vector<std::complex<T>> tmpbuf(nfft_h);
+            kiss_fft_stride(plan.substate, (const std::complex<T>*)timedata, &tmpbuf[0], 1);
+            /* The real part of the DC element of the frequency spectrum in st->tmpbuf
+             * contains the sum of the even-numbered elements of the input time sequence
+             * The imag part is the sum of the odd-numbered elements
+             *
+             * The sum of tdc.r and tdc.i is the sum of the input time sequence.
+             *      yielding DC of input time sequence
+             * The difference of tdc.r - tdc.i is the sum of the input (dot product) [1,-1,1,-1...
+             *      yielding Nyquist bin of input time sequence
+             */
+
+            freqdata[0]         = std::complex<T>(tmpbuf[0].real() + tmpbuf[0].imag(), 0);
+            freqdata[nfft_h]    = std::complex<T>(tmpbuf[0].real() - tmpbuf[0].imag(), 0);
+
+            static constexpr T half = 0.5;
+            
+            for (int k = 1 ; k <= nfft_h / 2 ; ++k)
+            {
+                auto fpk  = tmpbuf[k];
+                auto fpnk = std::conj(tmpbuf[nfft_h-k]);
+                auto f1k = fpk + fpnk;
+                auto f2k = fpk - fpnk;
+                auto tw  = f2k * plan.super_twiddles[k-1];
+                freqdata[k]         = half * (f1k + tw);
+                freqdata[nfft_h-k]  = half * std::conj(f1k - tw);
+            }
+        }
+
+        template<typename T>
+        void kiss_fftri(const kiss_fftr_state<T>& plan, const std::complex<T>* freqdata, T* timedata)
+        {
+            if (!plan.substate.inverse)
+                throw std::runtime_error("bad fftr plan : need an inverse plan. This is a forward plan");
```

Comment:
This kind of thing should be a `DLIB_ASSERT`.

Reply:
yep